### ISSUE TITLE
Going back doesn't work on pdf.js page

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#1218]
When opening a PDF with pdf.js, it seems no new history entry is generated for the pdfjs viewer, so going back doesn't take us back to the page the PDF was opened from.

cc @Kingdread

**Comment by Kingdread** [comment]
That's strange, seems to work sometimes and sometimes not... Works fine when opening PDFs directly from duckduckgo's result page (`:back` does take you back to the result page), doesn't work e.g. on http://stlab.adobe.com/wiki/index.php/File:Test.pdf (the "Test.pdf" link, goes back to the page before the adobe wiki or says "At beginning of history." if you started directly at the adobe page)

**Comment by Kingdread** [comment]
Actually it seems to have something to do with redirects. duckduckgo links to a redirect URL (just like Google), which does get its own history entry, so going back there works. A direct link does not generate such an entry, so there's really no new history entry created.

**Comment by lahwaacz** [comment]
Note that as a result, PDFs are not saved in the session.

**Comment by nabsiddiqui** [comment]
Is there a way for qutebrowser to determine if a pdf is being opened and save the previous history before it opens?

**Comment by nabsiddiqui** [comment]
Is there a way for qutebrowser to determine if a pdf is being opened and save the previous history before it opens?

**Comment by Kingdread** [comment]
Yes, there is a way to detect if a pdf is being opened, we already do it because we have to manually load pdf.js. The problem is that `QWebFrame.setContent()` (which we use to set the content to pdf.js) does not create a new history entry, which means the entry of the page where the pdf link was clicked is used. The API does not seem to provide a way to change that. So if you click a PDF link on "http://foo.bar", your history stays "http://foo.bar", and going back will tell you that there's no previous page. So saving the history probably wouldn't help, as the history is fine.

You could probably work around that by redirecting to some `qute://` page and provide pdf.js there (though you'd probably see how it plays with cross-domain requests), but before starting workarounds, it's probably best to wait for QWebEngine:

> [...] there are plans of enabling PDFium (the Chromium PDF viewer) in QtWebEngine, so I guess we wouldn't need pdfjs at all there.

**Comment by nabsiddiqui** [comment]
OK thanks for the clarification.

If I am understanding correctly, why not just save the current page's url to the history before the PDF is loaded?  While it is technically true that the user hasn't gone to a new page, it would allow users to go backwards from the PDF.

**Comment by lahwaacz** [comment]
@Kingdread You could try to duplicate the current history entry with some Javascript magic before loading pdf.js. E.g. I think something like this might work:

```js
history.pushState(history.state, "", location.href);
```

**Comment by nabsiddiqui** [comment]
@lahwaacz yeah this more of what I was trying to convey in my original question. I just worded it wrong.

**Comment by lahwaacz** [comment]
Yeah, you beat me with your post by a couple of minutes :laughing:

**Comment by Kingdread** [comment]
Unfortunately I'm not a wizard :wink:

I've tried various javascript based approaches, but none of them gave the right result. I'm no javascript expert, but as far as I can tell, `pushState` does not create a *real* history entry, and navigating between pushed states only sends some event and does not actually reload the page (which would kind of defeat the purpose of `pushState` in the first place). I don't remember exactly what went wrong, but IIRC I couldn't fix it with a simple `pushState`, though as I said, I'm not a javascript-developer, so if you feel like giving it a try, feel free to do so :wink:

At the very least you need a `window.onpopstate` handler, and I'm not sure what didn't work (I've since removed the experiment).

**Comment by lahwaacz** [comment]
Hmm, unfortunately me neither (both wizard and javascript expert)...

Actually, the behaviour is now different than before - going back from the PDF now actually works, but if I have say 3 pages in the tab's history and keep going backwards and forwards long enough, I end up with all three items rewritten to the PDF's URL. That is with pdfjs 1.4.20 and qutebrowser 0.8.2...

By the way, there is some [serious history rewriting](https://github.com/mozilla/pdf.js/blob/9c95d089dea692f2ab4cd4c8b188555d2550b27f/web/pdf_history.js#L201-L218) in pdfjs so I guess somehow it doesn't play nice with qtwebkit/qutebrowser.

**Comment by lahwaacz** [comment]
Today I came up with this code, which adds the missing history entry for direct links. The problem is that going back from the PDF does not automatically reload the page, likewise going forward again. The rest of the history (i.e. HTML <-> HTML) is not affected.

And I still have the problem with redirecting links...

```diff
diff --git a/qutebrowser/browser/webkit/webpage.py b/qutebrowser/browser/webkit/webpage.py
index a3d2a01..996418c 100644
--- a/qutebrowser/browser/webkit/webpage.py
+++ b/qutebrowser/browser/webkit/webpage.py
@@ -226,8 +226,10 @@ class BrowserPage(QWebPage):
         except pdfjs.PDFJSNotFound:
             page = jinja.render('no_pdfjs.html',
                                 url=reply.url().toDisplayString())
-        self.mainFrame().setContent(page.encode('utf-8'), 'text/html',
-                                    reply.url())
+        frame = self.mainFrame()
+        if reply.url() not in [item.url() for item in self.history().items()]:
+            frame.evaluateJavaScript("location.href = location.href + '#'")
+        frame.setContent(page.encode('utf-8'), 'text/html', reply.url())
         reply.deleteLater()
 
     def shutdown(self):
```

**Comment by The-Compiler** [comment]
If we could fix this somehow (maybe even by manipulating the history from Python), it'd be nice to have this directly in the browsertab API so it works everywhere `setContent` is used.

OTOH, couldn't we just avoid using `setContent` entirely and load `qute://pdfjs?url=...` or something?

**Comment by lahwaacz** [comment]
> OTOH, couldn't we just avoid using `setContent` entirely and load `qute://pdfjs?url=...` or something?

That's probably the only way to make navigating history 100% reliable, but unfortunately pdfjs can load PDF files only from the same domain, i.e. `qute://pdfjs` here. That's basically what @Kingdread said in https://github.com/The-Compiler/qutebrowser/issues/1218#issuecomment-237634967, but I actually have the code for that :wink:

**Comment by The-Compiler** [comment]
Fixed with the PDF.js rewrite which just landed in master!
